syntax = "proto3";

package library;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

// Service to manage the library
service Library {
  // Book RPCs
  rpc CreateBook(CreateBookRequest) returns (Book);
  rpc GetBook(GetBookRequest) returns (Book);
  rpc ListBooks(ListBooksRequest) returns (ListBooksResponse);
  rpc UpdateBook(UpdateBookRequest) returns (Book);
  rpc DeleteBook(DeleteBookRequest) returns (google.protobuf.Empty);

  // Member RPCs
  rpc CreateMember(CreateMemberRequest) returns (Member);
  rpc GetMember(GetMemberRequest) returns (Member);
  rpc ListMembers(ListMembersRequest) returns (ListMembersResponse);
  rpc UpdateMember(UpdateMemberRequest) returns (Member);
  rpc DeleteMember(DeleteMemberRequest) returns (google.protobuf.Empty);

  // Borrowing RPCs
  rpc BorrowBook(BorrowBookRequest) returns (Borrowing);
  rpc ReturnBook(ReturnBookRequest) returns (Borrowing);
  rpc ListBorrowedBooks(ListBorrowedBooksRequest) returns (ListBorrowedBooksResponse);
}

// Book message
message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  int32 published_year = 5;
  int32 quantity = 6;
  int32 quantity_available = 7;
}

// Member message
message Member {
  string id = 1;
  string name = 2;
  string email = 3;
  string phone = 4;
}

// Borrowing message
message Borrowing {
    string id = 1;
    string book_id = 2;
    string member_id = 3;
    google.protobuf.Timestamp borrow_date = 4;
    google.protobuf.Timestamp return_date = 5;
    google.protobuf.Timestamp due_date = 6;
}


// --- Request and Response Messages ---

// Book messages
message CreateBookRequest {
  string title = 1;
  string author = 2;
  string isbn = 3;
  int32 published_year = 4;
  int32 quantity = 5;
}

message GetBookRequest {
  string id = 1;
}

message ListBooksRequest {
  // No parameters for now, could add pagination later
}

message ListBooksResponse {
  repeated Book books = 1;
}

message UpdateBookRequest {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  int32 published_year = 5;
}

message DeleteBookRequest {
  string id = 1;
}

// Member messages
message CreateMemberRequest {
  string name = 1;
  string email = 2;
  string phone = 3;
}

message GetMemberRequest {
  string id = 1;
}

message ListMembersRequest {
  // No parameters for now, could add pagination later
}

message ListMembersResponse {
  repeated Member members = 1;
}

message UpdateMemberRequest {
  string id = 1;
  string name = 2;
  string email = 3;
  string phone = 4;
}

message DeleteMemberRequest {
  string id = 1;
}

// Borrowing messages
message BorrowBookRequest {
  string book_id = 1;
  string member_id = 2;
}

message ReturnBookRequest {
  string book_id = 1;
  string member_id = 2;
}

message ListBorrowedBooksRequest {
  string member_id = 1;
}

message ListBorrowedBooksResponse {
    repeated Borrowing borrowings = 1;
}
